### ISSUE TITLE
`pip compile --upgrade` behavior with yanked packages

### ISSUE THREAD

**Issue by zanieb** [astral-sh/uv#3644]
Prompted at https://github.com/astral-sh/uv/issues/3602#issuecomment-2115761269

When a yanked package is pinned in a lockfile and `pip compile --upgrade` is used and there is no new version of the package in the input range, should we

1. Downgrade to the unyanked version
2. Error that a yanked version is being used
3. Warn that a yanked version is being used
4. Silently continue to use the yanked version

Note the yanked package is _not_ pinned in the input requirements.

**Comment by notatallshaw** [comment]
Thinking out loud, it seems like the most likely scenario this would happen in is if the compile originally ran and the package was not yanked, and then it was run at a later time and the package was yanked.

Given this is in the case where the yanked package is not pinned in the requirements, I would personally expect "1." with maybe a warning.

**Comment by zanieb** [comment]
Yeah that's my feeling as well. I would find it weird that `--upgrade` downgrades a package but 🤷‍♀️ yanks are rare.

**Comment by notatallshaw** [comment]
Does `--upgrade` normally guarantee the versions pinned in the old output file will be lower bounds of the new output file?

For example if I had a direct dependency on `a`, and when I ran pip compile for the first time it created a pinning `b==2.0`, but then when I ran upgrade the next day the newer version of `a` now requires `b==1.0`, would `a` be upgraded and `b` downgraded or would `a` now never be able to upgrade?

I'm not that familiar with the behavior of `pip compile --upgrade` but I know `conda upgrade --all` will downgrade packages.

**Comment by konstin** [comment]
I'd definitely want a warning, i've been using a broken release. Option 1 seems to be the most helpful behavior to me.

**Comment by zanieb** [comment]
The next step is to verify our existing behavior, if anyone is interested.

**Comment by notatallshaw** [comment]
FYI, the existing behavior is currently "1." (but with no warning), I recently came across this with matplotlib which yanked 3.9.1: https://pypi.org/project/matplotlib/3.9.1/

My output file had `matplotlib==3.9.1` in it, and when I reran `uv pip compile --upgrade` I got  `matplotlib==3.9.0` .
